src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -365,12 +365,15 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
               annotation()
             case _ if in.token == LBRACE =>
               atPos(in.pos) {
-                val elts = inBracesOrNil(commaSeparated(annVal()))
+                val elts = inBracesOrNil(if (in.token == RBRACE) Nil else commaSeparated(annVal()))
                 if (elts.exists(_.isEmpty)) EmptyTree
                 else Apply(ArrayModule_overloadedApply, elts: _*)
               }
             case _ if in.token == IDENTIFIER =>
               qualId(orClassLiteral = true)
+            case _ =>
+              syntaxError(in.currentPos, "Annotation value expected", skipIt = true)
```

Comment:
I think following our reasoning on how non-constant values weren't supported before and aren't supported still, we should continue to just skip those, no?  No error?

Reply:
Yeah, the syntax allows a lot more https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-ElementValue

For example, this should not give a syntax error

```
public class A {
  public @interface Ann { int value(); } 
  @Ann((1 == 2) ? 3 : 4)
  public static void foo() { return; }
}
```

Reply:
I'll amend and add more neg tests. The existing bailout mechanism doesn't deal with the unsupported syntax within array initializers.

Reply:
I also found an existing bug:

```
@Ann(values = {"foo", "bar"})
class Test {}

@interface Ann {
    String[] values();
}
```

```
sandbox/Test.java:2: error: `package' expected but `class' found.
class Test {}
^
sandbox/Test.java:6: error: identifier expected but eof found.
}
 ^
two errors found
```

Another edge case arises when trailing commas are uses in array initializers:

```java
package p1;

@Ann(values = {1, 2, })
class Test {}

@interface Ann {
    int[] values();
}
```

That's legal but gives a `MatchError` in 2.12.x (like the main test case in this PR). My current patch drops the arguments unnecessarily. I'll refine it to support the trailing comma.